feat(radio): refactor of switch handling to better support customisable switches

philmoz opened the pull request:
This started out as a way to let users set the LED color for the SA & SD switches on the GX12 (see PR #6053); but it became clear that we really needed to be able to have these as fully customisable switches (like SW1 - SW6).

This PR is a big refactor of the switch handling to achieve this.
All momentary switches with LED's can be set as customisable switches. Switch name ordering is maintained (SA & SD come before SW1-6).

The customisable switch settings in the model yaml file are now readable instead of being magic numbers.

I plan to extend this by also adding configuration of the customisable switches in the radio settings.
Like we have with the enabled features, the custom switch settings in the model can be set to use the radio settings.
So you will be able to have radio level settings for all customisable switches; but still override per model.

Companion still needs a lot of work. Basic read/write of the yaml files is done; but the model edit, radio edit and simulator will need fixing (@elecpower - I will need your help here).

WARNING: backup your radio and model files before playing with this. There are changes to the yaml files that are not backward compatible.

I have tested on GX12, T15, T20V2, Bumblebee and ST16.
T20, TPro, and TProV2 are untested as I don't have these radios.

Todo:
- [x] Companion support.
- [x] Radio settings for customisable switches.
- [x] Test radios with flex switches.
- [x] Investigate if flex switches can be unified.

philmoz commented:
> It will be a few days before I even think of looking at this, but I can do the tests for T20, TPro, and TProV2. I expect the T20 would be fine as it is somewhat similar to the T20V2 in this regard. I take it #6053 will be updated on the back of this?

No rush - I have more testing to do.

This supersedes 6053.

elecpower commented:
When the radio side is all but finished I'll tackle the Companion side. Working on the  build toolchain at present.

philmoz commented:
I have added setup for custom switches to the radio hardware page (and radio data structure) so you can define type, name and startup position at the radio level.

![screenshot_gx12_25-04-23_08-08-51](https://github.com/user-attachments/assets/3380d3dc-3b53-4c8b-8d44-9288903a6fca)

Added a new type for model custom switch setup so you can choose to use the radio settings for each switch.

![screenshot_gx12_25-04-23_08-13-28](https://github.com/user-attachments/assets/56703793-e98f-41d9-bc7c-9f2ce7f9becb)

Note: there is no group setting for custom switches at the radio level. I thought it might be too confusing.

philmoz commented:
@elecpower I've got yaml read/write working for companion with the changes

philmoz commented:
I think the core changes are done for the firmware (pending any bug fixes).

@elecpower for Companion I have yaml read / write working and have updated the model & radio edit to include the changes. I have added the custom switches to the simulator UI; but have not wired them to the running firmware.
Todo for Companion:
- [ ] Edit custom switch colors on the model setup edit page
- [ ] Edit custom switch colors on the radio hardware edit page
- [ ] Get the simulated switch type info for the simulator from the model or radio settings
- [ ] Link the simulated switches to the running firmware in the simulator. This needs to be two way as the switch state can be changed by the user or by the firmware.
 
Any suggestions would be appreciated.

<img width="530" alt="Screenshot 2025-04-23 at 8 17 01 PM" src="https://github.com/user-attachments/assets/f593f271-e41b-414c-bd8d-ae6ae0ee795f" />

<img width="769" alt="Screenshot 2025-04-23 at 8 17 18 PM" src="https://github.com/user-attachments/assets/270e5991-5cc8-4954-b610-e47f9e4d1e13" />

<img width="1171" alt="Screenshot 2025-04-23 at 8 16 49 PM" src="https://github.com/user-attachments/assets/80ba6302-57fd-42f2-9447-a324d84b425b" />

elecpower commented:
@philmoz thanks for the update. I'll start to have a look in the next few days when free time.

Use QColorDialog along the lines used for simulated radio case for the colour picker in Radio Profile settings.

elecpower commented:
Looks like we need to create new simulator widgets for six position buttons/crows feet with LED colours as an alternative to 6 position knob and LED colours for all switches (not just current SA/SD)

elecpower commented:
These are likely on your TODO list but just in case. Some initial feedback from testing.

GX12 libsim:
- after creating a new model, switches SW1-6 not initialised to Global type

Companion GX12 profile
- after creating a new model, all customisable switches should have Global type
- model setup switch warnings missing for A & D
- switch dropdown lists broken for A & D
- setting hardware SW1-6 to None and Global for model the switch dropdown lists contain the None switch

philmoz commented:
The default for SW1-6 should be model setup, 2POS and group 1 for all of them to mimic the current way it works.

elecpower commented:
> The default for SW1-6 should be model setup, 2POS and group 1 for all of them to mimic the current way it works.

I can understand not changing existing models but IMO is counter intuitive especially since defaults are inconsistent across all customisable switches. How would we treat future customisable physical switches?

elecpower commented:
Setting new models to Global does not break anything

philmoz commented:
I've changed the default type and fixed the issues.

elecpower commented:
Retested after your last commit and all appear fixed except:
- model setup switch warning choice should be hidden in Companion and radio for ALL switches of type Toggle. Currently only non-customised Toggles are hidden.

philmoz commented:
I've had another go at fixing the custom switch logic and switch warnings.

elecpower commented:
Retested and switch warning visibility changes to reflect global and model level type.

Switch warnings are enabled by default however when a switch is changed from None or Toggle at global or model, thus becoming visible, the corresponding checkbox is not checked.

philmoz commented:
If a switch is not available for warning the corresponding warning flag is cleared (so the checkbox will not be set). If the availability changes this does not change the flag state so the checkbox remains unchecked.

elecpower commented:
IMO a switch becoming available should attract the default state of warning enabled.
The state for all switches on model creation was changed some time ago from disabled to enabled due to user feedback.
My two cents worth.

philmoz commented:
> IMO a switch becoming available should attract the default state of warning enabled. The state for all switches on model creation was changed some time ago from disabled to enabled due to user feedback. My two cents worth.

That's not what is happening in companion for physical switches.
E.G. on TX16S set SA to warning enabled (position does not matter). Change radio setting for SA to None. Open model settings - SA is not shown in warning. Change radio setting back to 3POS. Open model setting and switch warning is disabled for SA.

Companion works very differently to the radio in this regard.
The radio does not alter the switch warning state when the switch type is changed.
Companion splits the warning state into two variables (position and enabled) and changes the enabled state when the model settings is opened.
Companion needs to be fixed to work the way the radio does.

elecpower commented:
B&W differed from colour in handling and Companion was stuck in the middle and likely ended up in the mess you fixed. Thanks.

philmoz commented:
I'm guessing a lot of the companion code has evolved over time ro handle difference and changes - a lot of it makes me scratch my head.

Have you considered a singleton for the radio settings to avoid passing the GeneralSettings& around everywhere?

3djc commented:
How do you deal with numbering, and already available scripts that loop to LED_STRIP_COUNT to create ambiance ? Those will affect off led too which is likely unwanted ?

philmoz commented:
> How do you deal with numbering, and already available scripts that loop to LED_STRIP_COUNT to create ambiance ? Those will affect off led too which is likely unwanted ?

How do you know it's unwanted - it's just as likely that this is exactly what the end user intended.

3djc commented:
Lets be realistic, there are already many apparence led scripts that mimic all kind of effects. It is quite unlikely that you want off functional leds to be part of those, or there needs to be at the very least a way to exclude those that doesn't require user to be a programmer. Many users use lua scripts,but only a handful actually author lua scripts

pfeerick commented:
Bringing us back to what I said right from the start... rather than **assume** we know what the user wants, perhaps there needs to be a specific "lua" mode. Off/Red/White/Blue/Lua. Once you **know** that is the intent, there is no longer a clash between the functional switch led function, and it being controlled by a lua script. :facepalm:

3djc commented:
Unfortunately, not quite true. You might have a cosmetic script (some radio in dev are even delivered with some running by default) but want to do something specific with one (or more) of the functional leds (like a 3 color arm like asked above). I'm personally in favour of two distinct lua call (or an option field on a single one)

pfeerick commented:
You'll have to run that by me again, as I don't quite know what you are talking about (but might also be because brain is not working by this time of night). I was specifically referring to the options where you define for a functional switch what its on and off colours are. If you define either to be Lua, Lua is what sets the colours. Better perhaps is a seperate Lua checkbox for both the on and off states? As then you can have set the on and off colours as per normal, and then you can also specify Lua can also control the colours.

3djc commented:
Ok, let me try again.

Most radio with cosmetics leds come with scripts that you run that loop from 0 to LED_STRIP_LENGHT and compute a color pattern for each led and set all the leds according to that pattern. Say you want a police like effect, you set all odd number led to red, all even to blue, and then next second, you do the opposite. That's for cosmetics

Now if you set your SW1 led to LUA, it will be reacting to the pattern above, when what you might want to do it something completely different.

pfeerick commented:
I have several radios that can do RGB led effects (including two with addressable gimbal rings), so know exactly what you mean there. 

So, to avoid Betaflight level of led mapping (and associated complexity), what are you proposing (or is this in addition to what I mentioned above)? Two different Lua RGB led SFs - one that would control all leds (any exceptions?), and one that would be constrained to only those (or single one?) chosen? And I guess one has priority over the other (or does it fall back to the order they are evaluated)?

3djc commented:
I would prefer a new call that only deals with fonctionnel leds, leaving current call to just cosmetic (and therefore all current scripts working).

Some users will want to use it for more advanced led control (like 3 Colors arming), they can do so without the risk of breaking the cosmetic leds effect.

Some users will want to include the functional leds to the effect, they can do so by editing the lua to add what is needed, or maybe later some scripts will offer the choice

If we have the two calls, I would think a setting for each led position to say LUA instead of a color would allow the right level of control (because you might just want to deal with say 3 Color arming for one custom switch, and leave etc controlling the others)

gagarinlg commented:
> I would prefer a new call that only deals with functional leds, leaving current call to just cosmetic (and therefore all current scripts working).

I agree!

3djc commented:
@philmoz i have made here a (untested) LUA part if you which to use it: https://github.com/EdgeTX/edgetx/commit/b990f9594940ebf6d3cc3d88ee57326a68f5d677

Currently, it allows LUA to control led which color for current state is set to 0

philmoz commented:
Latest changes:
- Lua function setRGBLedColor is limited to bling LEDs
- new Lua function setCFSLedColor can be used to override function switch led color
- Added a 'Lua override' toggle to all function switch colors to enable Lua override of switch color

The setCFSLedColor function can be called with either 1 or 4 arguments
- setCFSLedColor("switch name", r, g, b) - sets the override color for the named function switch
- setCFSLedColor("switch name) - removes the override color for the named switch and reverts the model/global setting colors.

3djc commented:
Tried it on GX12 sims, could not find where user can set the LUA override autorisation.

What are the CFS function switch settings in SYS/Hardware, are they like defaults ? I have a hard time (maybe because I haven't had breakfast yet !) to understand how it articulates with model settings. If you say in hardware SW1 starts UP for example, and in model set it part of a group where say SW2 starts UP, what happens ?

philmoz commented:
> Tried it on GX12 sims, could not find where user can set the LUA override autorisation.

Model & Radio settings for function switches:
![snapshot_01](https://github.com/user-attachments/assets/e0c3a894-b686-4c7f-8449-6d59279eb47b)
 
> What are the CFS function switch settings in SYS/Hardware, are they like defaults ? I have a hard time (maybe because I haven't had breakfast yet !) to understand how it articulates with model settings. If you say in hardware SW1 starts UP for example, and in model set it part of a group where say SW2 starts UP, what happens ?

Radio settings are only used if model setting is set to 'Global' for switch.
![snapshot_04](https://github.com/user-attachments/assets/07614258-f14c-4e33-b662-0723fec69d47)

3djc commented:
Cool, unsure why I wasn't getting that, a clean dir compile show it now :). Maybe "Allow Lua override" would be a better wording ? But that sounds good.

I'm still a bit puzzled about the practicality of the global/model settings for those, as the global is really meant for standalone (non grouped) switches.

philmoz commented:
> I'm still a bit puzzled about the practicality of the global/model settings for those, as the global is really meant for standalone (non grouped) switches.

For cases where you just want to use the switch the same way for all/most models and don't need grouping.
E.G. I use SA on GX12 for arming with LED color set to red.

3djc commented:
Yes, SA/SD on gx12 are the perfect example where it makes a lot of sense

pfeerick commented:
I've not tried this yet, but from the screenshots and updates it sound like it is coming together well. Am I correct in understanding that this won't need any changes to the RGBLed SF, as the decision on whether to override the customisable switches will be made as part of the customisable switch setting ("Lua Override" => "Allow Lua Override") ... as well as by which call the Lua author uses? If so, sounds perfect, and will just need to update and cross-reference in the manual accordingly once this is in. 

And being able to have a global setting for any of the customisable switches makes sense, not *just* the GX12 SA/SD... i.e. if you fly a bunch of things with flight controllers (quad, plane, car, whatever), and want to set the switches up as modes, you can do it globally, once, and then have model level override for the odd model/thing you want different.

philmoz commented:
@elecpower - I added buttons to the model setup window for editing the on/off switch colors. These work except when I close the QColorDialog window focus switches back to the main companion window instead of the model setup window. Any ideas on how to fix this?

elecpower commented:
I haven't checked your code but as a quick response check the parent object pointer passed to the colour dialog is the model edit dialog. If not that, then I'll have to dig.

philmoz commented:
I've tried various combinations (nulll, this, parent, parent->parentWidget(), etc) for the parent value in both the QColorDialog constructor and the getColor function but nothing seems to work. Maybe a MacOS issue?

elecpower commented:
@philmoz focus returns to the model edit for Ubuntu. I haven't tested Windows as my MSYS environment is set up for Qt 6.
In the Lambda functions you could try adding this->setFocus() as the last line though doubt it will work as Qt thinks it knows best.
Another suggestion is to go old school and not use Lambdas (I do prefer as you have coded). Sometimes it the last resort solution.
Also the emit updateDataModels() calls in the colour Lambdas are unnecessary in this instance and add lag.

philmoz commented:
@elecpower I have added a UI for the function switches and connected it to the simulator. When you have time can you take a look please - the button UI could use some work.

<img width="733" alt="Screenshot 2025-07-09 at 9 23 48 pm" src="https://github.com/user-attachments/assets/ce3878ff-a96c-41da-8b3f-901fe49f46cc" />

elecpower commented:
...used GX12 for test

philmoz commented:
> Just a very quick test and the custom switches other than SA do not initialise to the colour to match the switch position.

The very first time you run the simulator with this PR it may behave oddly because there are saved state values for the switches that don't match what is now expected.

If you close and run again does it work better?

I thought I fixed this; but obviously not.

elecpower commented:
> If you close and run again does it work better?

I'll retest when I have some time

elecpower commented:
> I think I've fixed the first run switch initialisation issue.

It appears fixed.

Also on a side note in Ubuntu the timer_queue is not always stopping and thus Companion hangs. I haven't found a pattern.

elecpower commented:
> Is that related to this PR?

I don't believe it is. Seems to be a lottery.

raphaelcoeffic commented on radio/util/hw_defs/switch_config.py:
```diff
@@ -15,15 +15,22 @@
+        "SB": {"default": "3POS",   "display": [0, 0]},
+        "SE": {"default": "3POS",   "display": [0, 1]},
+        "SG": {"default": "TOGGLE", "display": [0, 2]},
         # right side
-        "SC": {"default": "3POS",   "display": [1, 1]},
-        "SD": {"default": "TOGGLE", "display": [1, 0]},
-        "SF": {"default": "3POS",   "display": [1, 2]},
-        "SH": {"default": "TOGGLE", "display": [1, 3]},
+        "SC": {"default": "3POS",   "display": [1, 0]},
+        "SD": {"default": "TOGGLE"},
+        "SF": {"default": "3POS",   "display": [1, 1]},
+        "SH": {"default": "TOGGLE", "display": [1, 2]},
```
Is this change intended? If yes, what would be the reason?

philmoz replied:
Yes. SA & SD are function switches - they have their own state display (LED) so don't need to be displayed on the LCD.

philmoz commented:
I've fixed the TProV2, MT12 and pre-start warning issues.

@elecpower Any thoughts on the color selection buttons UI?

elecpower commented:
<img width="759" height="142" alt="Screenshot from 2025-07-27 21-12-12" src="https://github.com/user-attachments/assets/3264824f-7f62-4341-8345-cb19063413c5" />

@pfeerick does that look better?
@philmoz I have a commit for this if you would like me to push

pfeerick commented:
Thanks @philmoz T-Pro V2 and MT12 look good now.

Outlines for the colour swatches looks great @elecpower ... thanks! :)

One final minor detail if either could check please... is it just me, or are the colour swatches on the model edit page not updating (i.e. to newly set colour) until you re-open the model edit screen (i.e. you need to close and re-open the model edit page to get the correct colours)? The radio setup screen is fine, but on Linux the model edit screen is playing up.

pfeerick commented:
Thanks, let's run with it then and see what happens 😁

On Mon, 28 July 2025, 10:46 am philmoz, ***@***.***> wrote:

> *philmoz* left a comment (EdgeTX/edgetx#6095)
> <https://github.com/EdgeTX/edgetx/pull/6095#issuecomment-3124859907>
>
> One final minor detail if either could check please... is it just me, or
> are the colour swatches on the model edit page not updating (i.e. to newly
> set colour) until you re-open the model edit screen (i.e. you need to close
> and re-open the model edit page to get the correct colours)? The radio
> setup screen is fine, but on Linux the model edit screen is playing up.
>
> Fixed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/6095#issuecomment-3124859907>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KM54CMVY3B6XPQO5FT3KVXH3AVCNFSM6AAAAAB3IE6FO2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMRUHA2TSOJQG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gagarinlg commented:
Yes and no

Yes, the latest nightly is a 3.0.0 preview, but this is the state since shortly after the 2.11 release.
Generally, the Nightlies are not well tested. You have to make a backup before updating and run extensive tests before flying with it.
Nightlies tend to be a bit unstable.
